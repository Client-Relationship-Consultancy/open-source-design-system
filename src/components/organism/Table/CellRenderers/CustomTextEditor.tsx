import React from "react";
import styled from "styled-components";

const TextEditor = styled.input`
  height: 100%;
  width: 100%;
  background-color: transparent;
  padding-left: 4px;
  border: none;

  :focus {
    font-size: 12px;
    outline: none;
    border-color: #719ece;
    border-radius: 2px;
    border-style: solid;
    border-width: 1px;
  }
`;

TextEditor.displayName = "TextEditor";

interface IProps {
  value?: string;
  maxLength?: number;
  useInnerRef?: boolean;
}

export default class CustomTextEditor extends React.Component<IProps> {
  componentRef = React.createRef<HTMLInputElement>();

  static defaultProps: Pick<IProps, "useInnerRef"> = {
    useInnerRef: false,
  };

  state = {
    value: this.props.value ?? "",
  };

  getValue() {
    return this.state.value;
  }

  handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ value: event.target.value });
  };

  // innerRef is deprecated in Styled Components v4. Uses ref instead.
  // This is to allow backward compatibility for older Styled Components.
  createRefProp = () =>
    this.props.useInnerRef ? { innerRef: this.componentRef } : { ref: this.componentRef };

  afterGuiAttached() {
    const input = this.componentRef.current;
    if (input) {
      input.focus();
      input.select();
    }
  }

  render() {
    return (
      <TextEditor
        {...this.createRefProp()}
        maxLength={this.props.maxLength}
        value={this.getValue()}
        onChange={this.handleChange}
        type="text"
      />
    );
  }
}
