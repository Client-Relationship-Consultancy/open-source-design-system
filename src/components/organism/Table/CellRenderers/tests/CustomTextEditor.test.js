import React from "react"
import { shallow } from "enzyme"
import "jest-styled-components"
import CustomTextEditor from "../CustomTextEditor"

describe("CustomTextEditor test", () => {
  let component
  beforeEach(() => {
    component = shallow(<CustomTextEditor value="mock initial value" />)
  })

  it("should match last snapshot", () => {
    expect(component).toMatchSnapshot()
  })

  it("sets the value to state using handleChange", () => {
    const newValue = "new mocked value"
    component.instance().handleChange({ target: { value: newValue } })
    expect(component.state().value).toEqual(newValue)
  })

  it("should use the ref prop when useInnerRef prop is false", () => {
    component.setProps({
      useInnerRef: false,
    })
    expect(component.instance().createRefProp()).toEqual({
      ref: component.instance().componentRef,
    })
  })

  it("should use the innerRef prop when useInnerRef prop is true", () => {
    component.setProps({
      useInnerRef: true,
    })
    expect(component.instance().createRefProp()).toEqual({
      innerRef: component.instance().componentRef,
    })
  })
})
