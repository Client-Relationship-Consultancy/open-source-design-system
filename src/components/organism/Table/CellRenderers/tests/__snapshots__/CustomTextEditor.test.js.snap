// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTextEditor test should match last snapshot 1`] = `
<TextEditor
  onChange={[Function]}
  type="text"
  value="mock initial value"
/>
`;
